src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -210,34 +201,61 @@ internal bool NullableSemanticAnalysisEnabled
 
-        internal bool ShouldRunNullableWalker
+        /// <summary>
+        /// Returns true if nullable analysis is enabled in this compilation.
+        /// </summary>
+        /// <return>
+        /// Returns true if analysis is explicitly enabled for all methods;
+        /// false if analysis is explicitly disabled; and otherwise returns true
+        /// if there are any nullable-enabled contexts in the compilation.
+        /// </return>
+        internal bool IsNullableAnalysisEnabled
         {
             get
             {
-                if (!_lazyShouldRunNullableWalker.HasValue())
+                if (!_lazyShouldRunNullableAnalysis.HasValue())
+                {
+                    _lazyShouldRunNullableAnalysis = (GetNullableAnalysisValue() ?? hasEnabledContexts()).ToThreeState();
+                }
+                return _lazyShouldRunNullableAnalysis.Value();
+
+                bool hasEnabledContexts()
                 {
                     if (Options.NullableContextOptions != NullableContextOptions.Disable)
                     {
-                        _lazyShouldRunNullableWalker = ThreeState.True;
                         return true;
                     }
-
-                    foreach (var syntaxTree in SyntaxTrees)
-                    {
-                        if (((CSharpSyntaxTree)syntaxTree).HasNullableEnables())
-                        {
-                            _lazyShouldRunNullableWalker = ThreeState.True;
-                            return true;
-                        }
-                    }
-
-                    _lazyShouldRunNullableWalker = ThreeState.False;
+                    return SyntaxTrees.Any(tree => ((CSharpSyntaxTree)tree).HasNullableEnables());
                 }
-
-                return _lazyShouldRunNullableWalker.Value();
             }
         }
 
+#if DEBUG
+        /// <summary>
+        /// Returns true if nullable analysis should be run and results ignored in cases where analysis
+        /// is not otherwise necessary. Used for testing in debug builds to increase the chance of catching
+        /// nullable regressions (e.g. https://github.com/dotnet/roslyn/issues/40136).
+        /// </summary>
+        /// <return>
+        /// Returns true unless analysis is explicitly disabled.
+        /// </return>
+        internal bool ShouldRunNullableAnalysisAndIgnoreResults => GetNullableAnalysisValue() != false;
+#endif
+
+        /// <summary>
+        /// Returns Feature("run-nullable-analysis") as a bool? value:
+        /// true for "always"; false for "never"; and null otherwise.
+        /// </summary>
+        private bool? GetNullableAnalysisValue()
+        {
+            return Feature("run-nullable-analysis") switch
+            {
+                "always" => true,
+                "never" => false,
+                _ => null,
```

Comment:
For unrecognized switch value, I think it's better to report an error to the user. That's already done for other options, e.g.:

https://github.com/dotnet/roslyn/blob/0845e86214cf3b084e17e9fcbdbd49ca6f804ca1/src/Compilers/CSharp/Portable/CSharpResources.resx#L5977-L5979

Reply:
> For unrecognized switch value, I think it's better to report an error to the user. That's already done for other options,

I don't think we hold feature flags to the same standard as regular compiler switches. That is why they are feature flags. It doesn't even look like we document how to set them.  

---
In reply to: [540926000](https://github.com/dotnet/roslyn/pull/49876#discussion_r540926000) [](ancestors = 540926000)